Features/task 1

kubmir opened the pull request.

Commit:
KA-226 Add NUnit framework and initial files

Commit:
KA-226 Initial items

Commit:
KA-226 Refactor initial notes

Commit:
KA-226 Add FindNotesAsync method with tests

Commit:
KA-226 Allow usage of C#7 in appveyor build

Commit:
KA-226 Add FindNoteById method with test

Commit:
KA-226 Add AddNewNoteAsync method with test

Commit:
KA-226 Add UpdateNoteAsync method with test

Commit:
KA-226 Add DeleteNoteByIdAsync method with test

Commit:
KA-226 Clean solution

Commit:
KA-226 Set up ApiVersioning with updated nuget packages

Commit:
KA-226 Fix definition of NoteEqualityComparer

Using Lazy to be created at the time when needed, private constructor
for creating new NoteEqualityComparer just in the NoteEqualityComparer class,
using instance and Instance to be thread safety.

Commit:
KA-226 Extend EqualityConstraint to using NoteEqualityComparer

Commit:
KA-226 Refactor Note

Commit:
KA-226 Remove NotesArray

Commit:
KA-226 Remove unnecessary logic from methods

Remove logic from methods to be just dummy methods, place static array of notes
directly into NoteController.

Commit:
KA-226 Refactor tests to work with modified controller

Commit:
KA-226 Refactor NoteController with routes

Commit:
KA-226 Clean solution for review

Commit:
KA-226 Rename NoteController to NotesController

PetrSvirak commented on NotesBackEnd/src/Notes.Api/App_Start/WebApiConfig.cs:
```diff
@@ -1,15 +1,23 @@
+                    ["apiVersion"] = typeof( ApiVersionRouteConstraint )
+                }
+            };
+            config.MapHttpAttributeRoutes(constraintResolver);
```
Since you use attribute routing, there should not be a default route - this way, an end-point will not work rather than work on a wrong URL.

PetrSvirak commented on NotesBackEnd/src/Notes.Api/Controllers/NotesController.cs:
```diff
@@ -0,0 +1,35 @@
+{
+    [ApiVersion("1.0")]
+    [Route("api/v{version:apiVersion}/notes/{id?}", Name = "Notes")]
+    public class NotesController : ApiController
```
Nice and clean controller 👍

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -0,0 +1,127 @@
+        [SetUp]
+        public void Init()
+        {
+            _controller = new NotesController();
```
Listen to your IDE 😄 (
![image](https://user-images.githubusercontent.com/15132110/33756940-2261ae38-dbf8-11e7-8f05-22df5cdef57e.png))

PS: do not forget about [line 28](https://github.com/kubmir/kentico-onboarding-cs/pull/2/files#diff-b15d71677aa42b8350280f2fec28c3d7R28)

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -0,0 +1,127 @@
+            HttpResponseMessage executedResponse = await response.ExecuteAsync(CancellationToken.None);
+            executedResponse.TryGetContentValue(out Note[] actualNotes);
+
+            Assert.Multiple(() =>
```
:+1: for the multiple asserts

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -0,0 +1,127 @@
+
+            IHttpActionResult response = await _controller.GetAsync(noteId);
+            HttpResponseMessage executedResponse = await response.ExecuteAsync(CancellationToken.None);
+            executedResponse.TryGetContentValue(out Note actualNote);
```
Since this repeats in every test, consider trying to extract a (generic) helper method out of the tests (if you fittingly simplify act (or any other part of test methods), you might increase readability).

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -0,0 +1,127 @@
+
+            Assert.Multiple(() =>
+            {
+                Assert.That(actualNotes.Length, Is.EqualTo(expectedNotes.Length));
```
You do not need to test a collection's length when you, later on, test it as a whole - the collection assertion will fail if there is an element missing on either side of it.

Pro tip: Always try to fail each test you write :)

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -0,0 +1,127 @@
+
+            IHttpActionResult response = await _controller.GetAsync();
+            HttpResponseMessage executedResponse = await response.ExecuteAsync(CancellationToken.None);
+            executedResponse.TryGetContentValue(out Note[] actualNotes);
```
Don't make the assumption on collection type - simply use most general (~ `IEnumerable<>`) type you can and let controller decide what collection fits it best - after all, nor you nor serialize minds… (of course, if the `array` was essential for a piece of code to work, I would not object at all)

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -0,0 +1,127 @@
+
+            Assert.Multiple(() =>
+            {
+                Assert.That(expectedNote, Is.EqualTo(actualNote).UsingNoteComparer());
```
Expected is supposed to be part of the builder rather than the first argument. 

Consider moving the expectation-describing expression to the arrange part of the test - it just an idea off top of my head, see how it looks and we can eventually discuss it (if it does not look like complete non-sense to you)

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Utils/EqualityConstraintExtension.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using NUnit.Framework.Constraints;
+
+namespace Notes.Api.Tests.Utils
```
Don't introduce JS' customs into C# 😛 It is about extending NUnit framework, or more precisely, its constraints system. What about `Constraints` folder with instead of `Utils`? Or if you more like the `Comparer` is the first citizen here, `Comparers` it would be (though to get to the comparer, one needs to go through the extension method)

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -0,0 +1,127 @@
+namespace Notes.Api.Tests.Controllers
+{
+    [TestFixture]
+    public class NotesControllerTest
```
Test classes can be internal.

PetrSvirak commented on NotesBackEnd/src/Notes.Api/App_Start/WebApiConfig.cs:
```diff
@@ -1,15 +1,23 @@
 
 namespace Notes.API
 {
     public static class WebApiConfig
```
It does not have to public class, make it internal.

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Utils/EqualityConstraintExtension.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace Notes.Api.Tests.Utils
+{
+    static class EqualityConstraintExtension
```
It is more about a custom, however, always state modifiers explicitly. For people from different than C# domain might not be apparent, that this class is internal.

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Utils/NoteEqualityComparer.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace Notes.Api.Tests.Utils
+{
+    class NoteEqualityComparer : IEqualityComparer<Note>
```
State the modifier explicitly here as well.

PetrSvirak replied:
Also, consider sealing the class when you expect nobody to extend it.

Note: It is a must with public API to decide whether or not a class can be derived as it adds to maintenance cost. In other code, sealing is more about expressing the purpose to other programmers - a little warning that given class is not meant for an extension (in the inheritance way of the meaning).

PetrSvirak commented on NotesBackEnd/src/Notes.Api/App_Start/WebApiConfig.cs:
```diff
@@ -1,15 +1,23 @@
+            {
+                ConstraintMap =
+                {
+                    ["apiVersion"] = typeof( ApiVersionRouteConstraint )
```
There should be no spaces around `ApiVersionRouteConstraint`

PetrSvirak reviewed (changes requested):
Good job, thanks. There are some minor things and some suggestions might find interesting - I believe it will be rather quick :). Just: Since some renaming will be required, be careful with git history.

Commit:
KA-226 Refactor WebApiConfig

Commit:
KA-226 Refactor NotesControllerTest

Commit:
KA-226 Make EqualityConstraintExtension internal explicitly

Commit:
KA-226 Make NoteEqualityComparer sealed and internal

Commit:
KA-226 Rename Utils to Comparers

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -7,28 +8,30 @@
+            };
+
             _controller.Configuration.Routes.MapHttpRoute(
                 name: "Notes",
```
It is a bit inconvenient that the `"Notes"` string can be found in 3 locations. Please, make it an internal constant in the `NotesController` for now (it would be `private` if was not needed in the tests or if the name itself (not that it is same everywhere) was important).

PetrSvirak commented on NotesBackEnd/tests/Notes.Api.Tests/Controllers/NotesControllerTest.cs:
```diff
@@ -113,15 +107,22 @@ public void Init()
             });
         }
+
+        internal async Task<(T ActualContent, HttpResponseMessage ResponseMessage)> GetExecutedResponse<T>(Func<Task<IHttpActionResult>> controllerFunction)
```
Really nice :-)
This method should be `private` tough, as it is used in this controller only.

PetrSvirak reviewed (dismissed):
It is really nice work. Thanks 👍 

A note regards commit message: Do not use `Refactor <something>` unless you describe what exactly was done further on in the message (go through [this article](https://chris.beams.io/posts/git-commit/) please). In some cases (such as adding explicit `internal` modifier), it might be a bit clearer to apply the fix globally and a name the commit after that. If multiple types of changes can be avoided in a single commit, it is usually more understandable… The reason is that reading history is easier and finding potential culprits is more straight-forward when one can imagine what each commit amended (especially if history examination appears a month or more after the original commits).

A note regards `Merge pull request` button below: While a PR was not approved or if checks did not pass, it would be great if the button was disabled. Have at [this article](https://help.github.com/articles/defining-the-mergeability-of-pull-requests/) and consider configuring your repository accordingly - it will push it yet to another level with CI :)

There are two tiny details that cause no immediate problems but should be fixed (see below) I don't mind them being fixed in next task, thus I approve this one - yet it is up to You. Merge the code at will.

Commit:
KA-226 Make help function GetExecutedResponse in tests private

Commit:
KA-226 Extract notes route to constant and use it

Commit:
KA-226 Fix bug with id parameter of PUT method

Commit:
Merge branch 'develop' into features/task-1

kubmir commented:
I fixed tiny details together with one bug. I also changed settings of the repository to have protected branches master and develop. 

Option "Dismiss stale pull request approvals when new commits are pushed" is checked so now your approval is necessary for PR not to be blocked.

PetrSvirak reviewed (approved):
Very well done 👍
